File: Interview Preparation Kit/Graphs/ConnectedCellGrid.js
// https://www.hackerrank.com/challenges/ctci-connected-cell-in-a-grid/problem
'use strict';

const fs = require('fs');

process.stdin.resume();
process.stdin.setEncoding('utf-8');

let inputString = '';
let currentLine = 0;

process.stdin.on('data', inputStdin => {
    inputString += inputStdin;
});

process.stdin.on('end', function() {
    inputString = inputString.replace(/\s*$/, '')
        .split('\n')
        .map(str => str.replace(/\s*$/, ''));

    main();
});

function readLine() {
    return inputString[currentLine++];
}

function search(grid, x, y) {
    if (x >= 0 && y >= 0 && x < grid.length && y < grid[0].length &&
        grid[x][y] == 1) {
        grid[x][y] = 0;
        return 1 + search(grid, x - 1, y) + search(grid, x + 1, y) +
                search(grid, x, y - 1) + search(grid, x, y + 1) +
                search(grid, x - 1, y - 1) + search(grid, x + 1, y - 1) +
                search(grid, x - 1, y + 1) + search(grid, x + 1, y + 1);
    }
    return 0;
}

// Complete the maxRegion function below.
function maxRegion(grid) {
    let largest = 0;

    for (let x = 0; x < grid.length; x++) {
        for (let y = 0; y < grid[0].length; y++) {
            if (grid[x][y] == 1) {
                let current = search(grid, x, y);
                if (current > largest) {
                    largest = current;
                }
            }
        }
    }

    return largest;
}

function main() {
    const ws = fs.createWriteStream(process.env.OUTPUT_PATH);

    const n = parseInt(readLine(), 10);

    const m = parseInt(readLine(), 10);

    let grid = Array(n);

    for (let i = 0; i < n; i++) {
        grid[i] = readLine().split(' ').map(gridTemp => parseInt(gridTemp, 10));
    }

    const res = maxRegion(grid);

    ws.write(res + '\n');

    ws.end();
}

File: Interview Preparation Kit/Graphs/ConnectedCellGrid.py
#!/bin/python3
# https://www.hackerrank.com/challenges/ctci-connected-cell-in-a-grid/problem
import math
import os
import random
import re
import sys

def search(grid, grid_rows, grid_cols, x, y):
    if (x >= 0 and y >= 0 and x < grid_rows and y < grid_cols
        and grid[x][y] == 1):
        grid[x][y] = 0
        return 1 + (search(grid, grid_rows, grid_cols, x - 1, y) +
            search(grid, grid_rows, grid_cols, x + 1, y) +
            search(grid, grid_rows, grid_cols, x, y - 1) +
            search(grid, grid_rows, grid_cols, x, y + 1) +
            search(grid, grid_rows, grid_cols, x - 1, y - 1) +
            search(grid, grid_rows, grid_cols, x + 1, y - 1) +
            search(grid, grid_rows, grid_cols, x - 1, y + 1) +
            search(grid, grid_rows, grid_cols, x + 1, y + 1))
    return 0

# Complete the maxRegion function below.
def maxRegion(grid):
    largest = 0

    grid_rows = len(grid)
    grid_cols = len(grid[0])
    for x in range(grid_rows):
        for y in range(grid_cols):
            if grid[x][y] == 1:
                current = search(grid, grid_rows, grid_cols, x, y)
                if current > largest:
                    largest = current

    return largest


if __name__ == '__main__':
    fptr = open(os.environ['OUTPUT_PATH'], 'w')

    n = int(input())

    m = int(input())

    grid = []

    for _ in range(n):
        grid.append(list(map(int, input().rstrip().split())))

    res = maxRegion(grid)

    fptr.write(str(res) + '\n')

    fptr.close()
